include/ttmlir/Dialect/TTNN/Utils/VirtualToPhysicalGrid.h

```diff
@@ -0,0 +1,76 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef TTMLIR_DIALECT_TTNN_UTILS_VIRTUALTOPHYSICALGRID_H
+#define TTMLIR_DIALECT_TTNN_UTILS_VIRTUALTOPHYSICALGRID_H
+
+#include "ttmlir/Dialect/TTNN/IR/TTNNOpsAttrs.h"
+
+#include "mlir/IR/AffineExpr.h"
+
+namespace mlir::tt::ttnn::utils {
+
+/// Creates an affine map that translates a virtual grid layout to a physical
+/// grid layout for a single device based on the specified tensor memory layout.
+///
+/// This function supports three types of tensor memory layouts:
+/// - WidthSharded: Maps a width-sharded virtual grid (1xN) to a physical grid
+///   with the specified shape.
+/// - HeightSharded: Maps a height-sharded virtual grid (Mx1) to a physical grid
+///   with the specified shape.
+/// - BlockSharded: Maps a block-sharded virtual grid (MxN) directly to a
+///   physical grid with the specified shape.
+///
+/// \param context The MLIR context.
+/// \param tensorMemoryLayout The tensor memory layout type.
+/// \param physicalGridShape The shape of the physical grid, defaults to {8, 8}.
+///
+/// \return An affine map that translates the virtual grid layout to the
+/// physical grid layout based on the specified tensor memory layout.
+auto SingleDeviceCreateVirtualToPhysicalLayoutMap(
```

Comment:
Please go thru the whole PR and fix accordingly:

https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html#using-auto-to-deduce-type